src/dune_engine/fs_memo.ml

```diff
@@ -2,10 +2,47 @@ open! Stdune
 open! Import
 open Memo.Build.O
 
+type t = { dune_file_watcher : Dune_file_watcher.t option }
+
+(* Ideally this should be an [Fdecl], but there are currently two reasons why
+   it's not:
+
+   - We read the workspace file before we start the inotify watcher, so [init]
+   is called after [t_ref] is used. This is clearly not good (it means we never
+   re-read the workspace file)
+
+   - There are tests that call [Scheduler.go] multiple times, therefore [init]
+   gets called multiple times. Since they don't use file watcher, it shouldn't
+   be a problem. *)
+let t_ref = ref { dune_file_watcher = None }
+
+let init ~dune_file_watcher =
+  match !t_ref.dune_file_watcher with
+  | Some _ ->
+    Code_error.raise
+      "Called [Fs_memo.init] a second time after a file watcher was already \
+       set up "
+      []
+  | None -> t_ref := { dune_file_watcher }
+
 (* Files and directories have non-overlapping sets of paths, so we can track
    them using the same memoization table. *)
 let memo =
-  Memo.create "fs_memo" ~input:(module Path) (fun _path -> Memo.Build.return ())
+  Memo.create "fs_memo"
+    ~input:(module Path)
+    (fun path ->
+      let { dune_file_watcher } = !t_ref in
+      Option.iter dune_file_watcher ~f:(fun dune_file_watcher ->
+          try Dune_file_watcher.add_watch dune_file_watcher path with
+          | Unix.Unix_error (ENOENT, _, _) -> (
+            (* If the file is absent, we need to wait for it to be created by
+               watching the parent. We still try to add a watch for the file
+               itself after that succeeds, in case the file was created already
+               before we started watching its parent. *)
+            Dune_file_watcher.add_watch dune_file_watcher (Path.parent_exn path);
+            try Dune_file_watcher.add_watch dune_file_watcher path with
+            | Unix.Unix_error (ENOENT, _, _) -> ()));
```

Comment:
In this case, shouldn't we proceed recursively subscribing to the parents until we find the one that does exist? Otherwise, we'll not be notified when `path` finally does get created.

Reply:
I just added a comment about this. Simply adding subscriptions won't be sufficient, so we need to assume that the parent dir exists.
We're not catching the exception coming from `add_watch` for the parent so if/when this situation arises then at least we'll know that this assumption was violated.

Reply:
Thanks!